### PR TITLE
[New Feature] Send Interactive List Messages

### PR DESCRIPTION
**Pull request by alienware**
WHAT
- [x] Adds `send_interactive_message` to `Api::Messages` which is aliased to `send_interactive_reply_buttons` and `send_interactive_list_messages`
- [x] `InteractiveAction` support `type` for `ReplyButton` and `ListMessage` with new fields for `ListMessage`
- [x]  Adds `InteractiveActionSection` and `InteractiveActionSectionRow` for `ListMessage` type

WHAT
- Partially completes #8 

CHANGELOG
- Adds `simplecov` to generate code coverage
- `Interactive` pushed the responsibility of validations to `InteractiveAction`
- `InteractiveActionButton` is renamed to `InteractiveActionReplyButton`
- `add_button` in `InteractiveAction` is renamed to `add_reply_button`

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -19,6 +19,7 @@ group(:test) do
   gem('rubocop-performance', require: false)
   gem('rubocop-sorbet', require: false)
   gem('webmock')
+  gem('simplecov', require: false)
```
Comment (ignacio-chiazzo) [inline Gemfile:22]: Could this be added to a separate PR, please?

File: lib/whatsapp_sdk/resource/error.rb [hunk 2]
```diff
@@ -21,10 +21,18 @@ def initialize(field, message)
         end
       end
 
+      class InvalidField < MissingValue; end
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/error.rb:24]: Where is it used?

File: lib/whatsapp_sdk/resource/interactive.rb [hunk 3]
```diff
@@ -85,15 +82,7 @@ def validate
...
-        buttons_count = action.buttons.length
-        raise WhatsappSdk::Resource::Error::InvalidInteractiveActionButton.new(
-          "invalid length #{buttons_count} for buttons in action. It should be 1, 2 or 3."
-        ) unless (REPLY_BUTTONS_MINIMUM..REPLY_BUTTONS_MAXIMUM).cover?(buttons_count)
-
-        button_ids = action.buttons.map(&:id)
-        raise WhatsappSdk::Resource::Error::InvalidInteractiveActionButton.new(
-          "duplicate ids #{button_ids} for buttons in action. They should be unique.",
-        ) unless button_ids.length.eql?(button_ids.uniq.length)
+        action.send(:validate)
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive.rb:85]: Could you make this method public and remove the `send`?

File: lib/whatsapp_sdk/resource/interactive_action.rb [hunk 4]
```diff
@@ -6,36 +6,120 @@ module Resource
...
+      sig { returns(T::Array[InteractiveActionReplyButton]) }
       attr_accessor :buttons
 
-      # TODO: attr_accessor :button
-      # TODO: attr_accessor :sections
+      # Returns the button of the Action. For list_message type, it's required.
+      #
+      # @returns button [String] .
+      sig { returns(String) }
+      attr_accessor :button
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action.rb:34]: why is it a `string`?

File: lib/whatsapp_sdk/resource/interactive_action.rb [hunk 5]
```diff
@@ -6,36 +6,120 @@ module Resource
     class InteractiveAction
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action.rb:6]: This class could use a test.

File: lib/whatsapp_sdk/resource/interactive_action.rb [hunk 6]
```diff
@@ -6,36 +6,120 @@ module Resource
...
+        when "reply_button"
+          json = { buttons: buttons.map(&:to_json) }
+        end
+
         json
       end
+
+      private
+
+      def validate
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action.rb:89]: let's make this public method since it's used outside.

File: lib/whatsapp_sdk/resource/interactive_action_section.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
+# typed: strict
+# frozen_string_literal: true
+
+module WhatsappSdk
+  module Resource
+    class InteractiveActionSection
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action_section.rb:6]: Could you add a test class for this?

File: lib/whatsapp_sdk/resource/interactive_action_section_row.rb [hunk 8]
```diff
@@ -0,0 +1,94 @@
+# typed: strict
+# frozen_string_literal: true
+
+module WhatsappSdk
+  module Resource
+    class InteractiveActionSectionRow
```
Comment (ignacio-chiazzo) [inline lib/whatsapp_sdk/resource/interactive_action_section_row.rb:6]: Same here, let's add a test class.

File: example.rb [hunk 9]
```diff
@@ -246,21 +246,21 @@ def print_message_sent(message_response)
   text: "I am the footer!",
 )
 
-interactive_action = WhatsappSdk::Resource::InteractiveAction.new()
+interactive_action = WhatsappSdk::Resource::InteractiveAction.new(
+  type: WhatsappSdk::Resource::InteractiveAction::Type::ReplyButton,
+)
```
Comment (ignacio-chiazzo) [inline example.rb:251]: Let's add it to the README file too

**Comment by ignacio-chiazzo** [comment]
I'm very excited about this feature. It's something I need and can't wait to use it. Great work!

**Comment by ignacio-chiazzo** [comment]
Hey @alienware could you rebase? 

I would like to cut a release in the next few days and include this PR. Thanks!

**Comment by alienware** [comment]
Sure @ignacio-chiazzo, I will get this ready in the next couple of days. Thank you for taking out time to review this.

File: lib/whatsapp_sdk/resource/error.rb (hunk 2 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/error.rb:24]: It should be used in `Component` class. I did add this error class but didn't refactor any existing code.

File: Gemfile (hunk 1 above)
Comment (alienware) [reply Gemfile:22]: Sure thing

File: lib/whatsapp_sdk/resource/interactive_action.rb (hunk 4 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action.rb:34]: This represents the `button` key of `interactive` object which is a text field only, as in, it does not have a `{ text: 'button' }` or `{ type: 'text', text: 'button' }` structure.

File: lib/whatsapp_sdk/resource/interactive_action_section_row.rb (hunk 8 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action_section_row.rb:6]: Done

File: lib/whatsapp_sdk/resource/interactive_action.rb (hunk 5 above)
Comment (alienware) [reply lib/whatsapp_sdk/resource/interactive_action.rb:6]: Done

**Review by ignacio-chiazzo — APPROVED** [review]

**Comment by ignacio-chiazzo** [comment]
Great work!
